output/mistral_fluentbit/mistral_fluentbit.c

```diff
@@ -0,0 +1,553 @@
+ * parameters. The stream used for error messages defaults to stderr but can
+ * be overridden here by setting plugin->error_log.
+ *
+ * Parameters:
+ *   plugin - A pointer to the plug-in information structure. This function
+ *            must set plugin->type before returning, if it doesn't the plug-in
+ *            will immediately shut down.
+ *   argc   - The number of entries in the argv array
+ *   argv   - A pointer to the argument array passed to main.
+ *
+ * Returns:
+ *   void - but see note about setting plugin->type above.
+ */
+void mistral_startup(mistral_plugin *plugin, int argc, char *argv[])
+{
+    /* Returning without setting plug-in type will cause a clean exit */
+
+    static const struct option options[] = {
+        {"error", required_argument, NULL, 'e'},
+        {"host", required_argument, NULL, 'h'},
+        {"port", required_argument, NULL, 'p'},
+        {0, 0, 0, 0},
+    };
+
+    const char *error_file = NULL;
+    const char *host = "localhost";
+    uint16_t port = 5170; /* This is the default TCP server port for Fluent Bit TCP connectivity */
+    int opt;
+    mode_t new_mode = 0;
+
+    /* This is thread safe because the receiving thread hasn't started at this point */
+    gettimeofday(&mistral_plugin_start, NULL);
+    timerclear(&mistral_plugin_end);
+
+    mistral_fluentbit_init(&fluentbit_tcp_ctx);
+
+    while ((opt = getopt_long(argc, argv, "e:h:p:v:", options, NULL)) != -1) {
+        switch (opt) {
+        case 'e':
+            error_file = optarg;
+            break;
+        case 'h':
+            host = optarg;
+            break;
+        case 'p': {
+            char *end = NULL;
+            unsigned long tmp_port = strtoul(optarg, &end, 10);
+            if (tmp_port == 0 || tmp_port > UINT16_MAX || !end || *end) {
+                mistral_err("Invalid port specified %s\n", optarg);
+                return;
+            }
+            port = (uint16_t)tmp_port;
+            break;
+        }
+        default:
+            usage(argv[0]);
+            return;
+        }
+    }
+
+    log_file_ptr = &(plugin->error_log);
+
+    if (error_file != NULL) {
+        if (new_mode > 0) {
+            mode_t old_mask = umask(00);
+            int fd = open(error_file, O_CREAT | O_WRONLY | O_APPEND, new_mode);
+            if (fd >= 0) {
+                plugin->error_log = fdopen(fd, "a");
+            }
+            umask(old_mask);
+        } else {
+            plugin->error_log = fopen(error_file, "a");
+        }
+
+        if (!plugin->error_log) {
+            plugin->error_log = stderr;
+            char buf[256];
+            mistral_err("Could not open error file %s: %s\n", error_file,
+                        strerror_r(errno, buf, sizeof buf));
+        }
+    }
+
+    /* Returning after this point indicates success */
+    plugin->type = OUTPUT_PLUGIN;
+
+    mistral_fluentbit_connect(&fluentbit_tcp_ctx, host, port, NULL, NULL);
```

Comment:
This connects to FluentBit after indicating to Mistral that everything is OK - this seems incorrect to me.

Reply:
I shall check `mistral_fluentbit_connect(...)` return value in the next release.

`mistral_fluentbit_connect(...)` starts a detached thread which implements the TCP connectivity.
Most of the error cases are handled later inside the thread itself. That's why it wasn't a priority to check what this function returns to the caller.